psc/src/main/java/com/pinterest/psc/metadata/client/PscMetadataClient.java

```diff
@@ -0,0 +1,172 @@
+ *
+ * This class is responsible for creating and managing the lifecycle of the backend-specific metadata clients,
+ * such as the {@link com.pinterest.psc.metadata.client.kafka.PscKafkaMetadataClient}, all of which should extend
+ * {@link com.pinterest.psc.metadata.client.PscBackendMetadataClient}.
+ *
+ * Each backend-specific metadata client is responsible for handling the actual metadata queries and operations.
+ *
+ * The creation of {@link PscBackendMetadataClient} relies on the supplied {@link TopicUri} at the time of
+ * method call to determine which backend-specific metadata client to create. The creation logic resides in
+ * the {@link PscMetadataClientCreatorManager} and the {@link PscBackendMetadataClientCreator} implementations.
+ * This model follows the same pattern as how backend clients are created in {@link com.pinterest.psc.consumer.PscConsumer}
+ * and {@link com.pinterest.psc.producer.PscProducer}.
+ *
+ * As such, each API method in this class will delegate to the appropriate backend-specific metadata client, and must
+ * accept a {@link TopicUri} to determine which backend-specific metadata client to use.
+ *
+ * The {@link TopicUri} supplied to each method does not need to be a full URI including the topic name. Instead,
+ * it should only include the URI prefix (up to the cluster name), since connecting to the cluster is the only
+ * information needed to determine which backend-specific metadata client to use. Supplying a full URI including the
+ * topic name will not cause any issues, but it is unnecessary.
+ */
+public class PscMetadataClient implements AutoCloseable {
+    private PscMetadataClientCreatorManager creatorManager;
+    private Environment environment;
+    private final PscConfigurationInternal pscConfigurationInternal;
+    private final Map<String, PscBackendMetadataClient> pscBackendMetadataClientByTopicUriPrefix = new ConcurrentHashMap<>();
+
+    public PscMetadataClient(PscConfiguration pscConfiguration) throws ConfigurationException {
+        this.pscConfigurationInternal = new PscConfigurationInternal(
+                pscConfiguration,
+                PscConfigurationInternal.PSC_CLIENT_TYPE_METADATA
+        );
+        initialize();
+    }
+
+    private void initialize() {
+        creatorManager = new PscMetadataClientCreatorManager();
+        environment = pscConfigurationInternal.getEnvironment();
+    }
+
+    /**
+     * List all the {@link TopicRn}'s in the cluster.
+     *
+     * @param clusterUri
+     * @param timeout
+     * @param timeUnit
+     * @return the list of {@link TopicRn}'s in the cluster
+     * @throws ExecutionException
+     * @throws InterruptedException
+     * @throws TimeoutException
+     */
+    public List<TopicRn> listTopicRns(TopicUri clusterUri, long timeout, TimeUnit timeUnit) throws ExecutionException, InterruptedException, TimeoutException {
```

Comment:
Just took a stab at this to see how much work would be involved. It is non-trivial. Introducing a `ClusterUri` class will require major refactoring of existing service discovery classes and their logic. The current logic in `TopicUri` and subclasses does not require a valid topic as the last component (surprisingly), and service discovery 

I tried two approaches - `ClusterUri extends BaseTopicUri` and `ClusterUri implements TopicUri`. Both of these approaches still required decent effort to refactor some commonly used classes such as service discovery, and both approaches introduces the risk of someone erroneously supplying a ClusterUri instance to the regular consumer/producer API's that require a TopicUri (since TopicUri is now a superclass of ClusterUri).

A third approach is to create separate Cluster* classes equivalent to TopicRn, TopicUri, BaseTopicUri, and KafkaTopicUri. This will still require major refactoring of service discovery and is more work.

I do think that we should refactor at some point to introduce the concept of ClusterUri, but given the effort I think the refactoring itself deserves a separate PR.

Reply:
Thanks for looking into this. If too much effort is involved, we can leave it in the backlog.
For `clusterUri` what is a sample input we'd be passing?